nts/Tag entity at Setup domain

StoyanGoranov opened the pull request:
Closes: https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
resolve conflicts and modify IdTag

StoyanGoranov committed:
include TagId property in Domain Setup Combination entity

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
create mock ITagWriter interface and TagWriterBehind

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
implement RfidTagForm in Setup

StoyanGoranov committed:
add RfidTagForm at CombinationForm and FormModel

https://github.com/Not-Endurance/not-timing-system/issues/224

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Combinations/CombinationForm.razor:
```diff
@@ -13,12 +14,14 @@
                Items="_athletes" @bind-Value="_model.Athlete" />
     <NotSelect T="Horse" Label="@Localizer.Get("Horse")" Placeholder="@Localizer.Get("Select a horse")"
                Items="_horses" @bind-Value="_model.Horse" />
+    <RfidTagForm @ref="_tagForm" CombinationNumber="_model.Number" />
```
Is it a form?

achobanov replied:
Also let's try to avoid `ref`. Isn't there a simpler way to get the tag given that it's not necessary to decouple this `RfidTagForm` component from `CombinationForm`.

achobanov replied:
Furthermore should this be rendered at all if StaticOptions aren't configured to use RFID?

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagForm.razor:
```diff
@@ -0,0 +1,50 @@
+            <MudFab Class="ml-2" Color="Color.Secondary" Size="Size.Small"
+            StartIcon="@Icons.Material.Filled.Delete" OnClick="RemoveTagHandler" />
+        </MudTooltip>
+    }
```
This conditional chain is not very well structured

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagForm.razor:
```diff
@@ -0,0 +1,50 @@
+    [Parameter]
+    public int CombinationNumber {get; set; }
+    public IdTag? Tag { get; private set; }
+    public bool _awaitingInput = false;
```
Would would be a better name that indicates the purpose of this field?

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagForm.razor:
```diff
@@ -0,0 +1,50 @@
+    else
+    {
+        <MudTooltip Text="Remove Tag">
+            <MudFab Class="ml-2" Color="Color.Secondary" Size="Size.Small"
```
Delete actions should be `Color.Danger` IMO. It is true that in the current theme the secondary color is also red, but I'm not a big fan of that and we'll probably end changing or customizing it. Using danger will always indicate that the action has consequences.

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagForm.razor:
```diff
@@ -0,0 +1,50 @@
+    public IdTag? Tag { get; private set; }
+    public bool _awaitingInput = false;
+
+    public async Task AddTagHandler()
+    {
+        _awaitingInput = true;
+        Tag = await _tagWriter.WriteTag(CombinationNumber);
+        _awaitingInput = false;
+    }
+
+    public void RemoveTagHandler()
```
Does `Handler` suffix provide anything meaningful in this context?

StoyanGoranov replied:
Thought it would be a nice convention to suffix methods that handle events with `Handler`. Anyway, removed the suffixes.

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagForm.razor:
```diff
@@ -0,0 +1,50 @@
+    @if (Tag == null)
+    {
+        <MudTooltip Text="Add Tag">
+            <MudFab Class="ml-2" Color="Color.Primary" Size="Size.Small" 
+            StartIcon="@Icons.Material.Filled.Add" OnClick="AddTagHandler" />
```
I Like the usage of `MudFab`. We should probably create `NotFabAdd` and ` NotFabDelete`

StoyanGoranov replied:
Using MudButton in the revised layout.

achobanov commented on src-v2/NTS.Domain.Setup/Entities/IdTag.cs:
```diff
@@ -2,19 +2,25 @@
+    //    Id = int.Parse(id);
+    //    Number = int.Parse(number);
+    //}
+    public IdTag(string tagId, int number)
```
Lets rename this to simply `Tag`

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagForm.razor:
```diff
@@ -0,0 +1,50 @@
+
+<MudContainer Class="d-flex flex-row">
+    @if(Tag!=null && !_awaitingInput){
+        <MudText Class="mt-2"> Contestant Number: @Tag.Number | Tag Id: @Tag.TagId</MudText>
```
Can't we use `ToString` here?

StoyanGoranov replied:
We could but I realized that the Tag number is redundant since it is configured by the same form.

achobanov replied:
`ToString` is called implicitly in the RenderTree method of every component.

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Ports/ITagWriter.cs:
```diff
@@ -0,0 +1,8 @@
+using NTS.Domain.Setup.Entities;
+
+namespace NTS.Judge.Blazor.Ports;
+public interface ITagWriter : ITransientService
```
Can be `ITagBehind`.

achobanov reviewed (changes requested):
Tag doesn't visualize in the combination form when already exists.

Additionally the feature doesn't look polished in the UI.

StoyanGoranov committed:
improve RfidTagForm layout

https://github.com/Not-Endurance/not-timing-system/issues/224

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagForm.razor:
```diff
@@ -0,0 +1,50 @@
+
+<MudText Class="mt-2" Typo="Typo.body2">Scan Rfid Chip (optional)</MudText>
+
+<MudContainer Class="d-flex flex-row">
```
I want to stress that you should use Mud components for layout instead of display: flex directly. `Class` properties should be used mostly for padding and margin.

StoyanGoranov replied:
This is too much stress 😄 Especially since I couldn't find a Mud Component that sets the flex direction or flex for that matter. Any suggestions?

achobanov replied:
See [MudStack](https://mudblazor.com/components/stack#basic-usage)

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagForm.razor:
```diff
@@ -0,0 +1,50 @@
+﻿@using NTS.Judge.Blazor.Ports
+@using NTS.Domain.Setup.Entities
+
+<MudText Class="mt-2" Typo="Typo.body2">Scan Rfid Chip (optional)</MudText>
```
Something is missing here

StoyanGoranov replied:
Not sure what are you referring to.

achobanov replied:
Localization

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagForm.razor:
```diff
@@ -0,0 +1,50 @@
+    }
+    @if (Tag == null)
+    {
+        <MudTooltip Text="Add Tag">
```
Something is missing here as well.

StoyanGoranov replied:
Dropped the use of Tooltip.

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagForm.razor:
```diff
@@ -0,0 +1,50 @@
+    }
+    else
+    {
+        <MudTooltip Text="Remove Tag">
```
And missing again

StoyanGoranov committed:
rename Models and Behinds

IdTag -> Tag
ITagWriter -> ITagBehind
TagWriterBehind -> TagBehind

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
rename RfidTagForm to RfidTagScanner and add Value param

*bonus* adds additional UI improvements

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
remove ref from RfidScanner and use new Value param at CombinationForm

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov commented:
Planning to make RfidTagScanner visible based on the StaticOptions with the introduction of the StaticOptions functionality in the Controllers PR.

StoyanGoranov committed:
improve naming of props and methods at RfidTagScanner

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
fix typo and use Tag.ToString instead of accessing properties

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
el fixo de ifo

StoyanGoranov committed:
inherit BindableValueComponent at RfidTagScanner

chosen this approach as the @bind-Value="_model.Tag"
states the component uses the _model.Tag property for two way binding

https://github.com/Not-Endurance/not-timing-system/issues/224

achobanov reviewed (commented):
Resolve left-over comments from before and it's all good

StoyanGoranov committed:
use MudStack instead of d-flex Html class

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
localize UI text at RfidTagScanner

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov commented:
Set a fixed height for the RfidTagScanner content and its buttons to ensure that the CombinationForm's height is not changed when MudProgress appears while awaiting chip scan.

StoyanGoranov committed:
add StaticOptions for DetectionConfiguration and use them at CombinationForm

available DetectionModes are Manual(default), Rfid and ComputerVision

https://github.com/Not-Endurance/not-timing-system/issues/224

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/Combinations/CombinationForm.razor:
```diff
@@ -14,7 +15,10 @@
     <NotSelect T="Horse" Label="@Localizer.Get("Horse")" Placeholder="@Localizer.Get("Select a horse")"
                Items="_horses" @bind-Value="_model.Horse" />
-    <RfidTagScanner CombinationNumber="_model.Number" Value="_model.Tag" />
+    @if (StaticOptions.RfidDetection())
```
Does `RfidDetection` imply bool value?

achobanov replied:
Still not there. A bool value should propose a question: `IsRfidDetectionEnabled`, `IsUsingRfid` are good suggestions

achobanov commented on src-v2/Judge/NTS.Judge.Blazor/Pages/Setup/RfidTags/RfidTagScanner.razor:
```diff
@@ -1,41 +1,43 @@
-               StartIcon="@Icons.Material.Filled.Delete" OnClick="RemoveTag">Remove Tag</MudButton>
+        if (Value == null)
+        {
+            <MudButton Style="height:32px" Class="mt-3" Color="Color.Primary" Size="Size.Small" Variant="Variant.Filled"
+                       StartIcon="@Icons.Material.Filled.Add" OnClick="AddTag">@Localizer.Get("Add Tag")</MudButton>
+        }
+        else
+        {
+            <MudPaper Style="height:32px" Class="mt-3">
+                <MudText Class="ml-2 mr-2" Typo="Typo.overline"> @Value </MudText>
+            </MudPaper>
+
+            <MudButton Style="height:32px" Class="ml-3 mt-3" Color="Color.Error" Size="Size.Small" Variant="Variant.Filled"
```
Why is this height style necessary? We don't want to use custom styles for layouts and colors, because it violates the consistency provided by the Mud theme.

achobanov commented on src-v2/NTS.Domain.Core/Configuration/DetectionConfiguration.cs:
```diff
@@ -0,0 +1,35 @@
+﻿namespace NTS.Domain.Core.Configuration;
+
+public class DetectionConfiguration
```
What's the purpose of this class?

StoyanGoranov committed:
integrate DetectionConfiguration class with StaticOptions

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
rename DetectionConfiguration.cs to DetectionMode

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
implement NotCollapse component

exposes Displayed property to enable the removal of collapsed elements
from the document flow; in addition component is suitable for future
extension with Rendered param to disable the rendering of content as
discussed at
https://stackoverflow.com/questions/68687308/is-there-any-way-to-disallow-a-component-from-rendering-using-shouldrender-ye

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
use NotCollapse in RfidScanTagger's spinner

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
StaticOptions improvementes

StoyanGoranov committed:
rename RfidTagScanner to RfidTagWriter

https://github.com/Not-Endurance/not-timing-system/issues/224

achobanov commented on src-v2/NTS.Domain.Core/Configuration/StaticOptions.cs:
```diff
@@ -10,7 +10,8 @@ public class StaticOptions : IStartupInitializer, ISingletonService
     static Model? _options;
     public static IRegionalConfiguration? RegionalConfiguration { get; set; }
-    public static DetectionConfiguration? DetectionConfiguration { get; set; }
+    public static bool RfidDetectionEnabled => _options != default && _options.DetectionMode == DetectionMode.Rfid;
+    public static bool VisionDetectionEnabled => _options != default &&  _options.DetectionMode == DetectionMode.ComputerVision;
```
Lets keep the API for StaticOptions consistent and use methods instead of dynamic properties.

achobanov commented on src-v2/Not/Not.Blazor/TM/NotCollapse.razor:
```diff
@@ -0,0 +1,24 @@
+    [Parameter]
+    public string Class { get; set; } = default!;
+    [Parameter]
+    public bool Expanded { get; set; }
+    [Parameter]
+    public bool Displayed { get; set; }
```
What's the idea of having both Displayed and Exapanded?

StoyanGoranov replied:
The point of the NotCollapse component is to remove the collapsed element from the document flow i.e stop elementс from taking screen real estate. Hence the Displayed property enables the removal.

StoyanGoranov committed:
use method instead of dynamic properties to check DetectionMode

https://github.com/Not-Endurance/not-timing-system/issues/224`

StoyanGoranov committed:
Merge branch 'develop' of https://github.com/Not-Endurance/not-timing-system into nts/Tag-entity

StoyanGoranov committed:
remove NotCollapse component

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
implement NotLabeledSpinner component

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
use LabelledSpinner in RfidTagWriter

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
add 'Is' to StaticOptions bool methods

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
fix wrong SafePattern method calls at Combination and Competition
Behinds

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
add empty list item to NotSelect

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
bold Type name in NotSelect empty message

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
remove NotLabelledSpinner and use NotifyHelper instead

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
revamp Spinner

https://github.com/Not-Endurance/not-timing-system/issues/224

StoyanGoranov committed:
add options param to prevent Dialog closure on backdrop clicks

https://github.com/Not-Endurance/not-timing-system/issues/224

achobanov commented on src-v2/Not/Not.Blazor/TM/Dialogs/NotDialogs.cs:
```diff
@@ -23,7 +23,8 @@ public NotDialogs(IDialogService mudDialogService, ILocalizer localizer)
     {
         var title = _localizer.Get("Create", " ", ReflectionHelper.GetName<T>());
-        var dialog = await _mudDialogService.ShowAsync<NotFormCreate<T, TForm>>(title);
+        var options = new DialogOptions { BackdropClick = false };
```
Do these options have to be created every time?

StoyanGoranov committed:
use DialogOptions as a property instead of a ctor var at NotDialog.cs

StoyanGoranov committed:
simplify NotSelect empty list message

StoyanGoranov committed:
layout fixes and improvements

https://github.com/Not-Endurance/not-timing-system/issues/224

achobanov commented on src-v2/Not/Not.Blazor/TM/Models/NotListModel.cs:
```diff
@@ -22,6 +22,11 @@ public static IEnumerable<NotListModel<T>> FromEntity<T>(IEnumerable<T> values)
         }
     }
+
+    public static NotListModel<T> GetEmptyValue<T>()
```
Is it getting an empty value or creating an empty list?

StoyanGoranov replied:
It does create an empty object with `Value=null;`

StoyanGoranov committed:
rename GetEmptyValue to CreateEmptyValue

https://github.com/Not-Endurance/not-timing-system/issues/224
